### PR TITLE
Pablo pallet: single asset deposit/withdrawal for constant product type of pools

### PR DESCRIPTION
**Pull request by saruman9**
## Issue

- [CU-2mkb3vt](https://app.clickup.com/t/2mkb3vt)
- [CU-2nnhwhf](https://app.clickup.com/t/2nnhwhf)
- [CU-2nnhx8y](https://app.clickup.com/t/2nnhx8y)
- [CU-2nnhxa9](https://app.clickup.com/t/2nnhxa9)
- [CU-2q94zpw](https://app.clickup.com/t/2q94zpw)
- [CU-2qevhzq](https://app.clickup.com/t/2qevhzq)
- [CU-2rh62qe](https://app.clickup.com/t/2rh62qe)
- [CU-2rpv7yc](https://app.clickup.com/t/2rpv7yc)
- [CU-2pcfnf3](https://app.clickup.com/t/2pcfnf3)
- [CU-2pj9but](https://app.clickup.com/t/2pj9but)
- [CU-2tkd28t](https://app.clickup.com/t/2tkd28t)

## Description

First attempt to implement single asset deposit (with fees) and withdrawal operations for Constant Product type of pools.

The development took place following Balancer's documentation:

- [website](https://docs.balancer.fi/v/v1/core-concepts/protocol/index#single-asset-deposit-withdrawal),
- [whitepaper](https://balancer.fi/whitepaper.pdf).

Please pay attention to `TODO` and `FIXME` comments (search by keyword `(TODO|FIXME)\(saruman9\)`).

I also made a few small stylistic corrections and fixes of minor mistakes (in the code and typo). If it was not permissible, then be free to inform me about it, I will revert these changes.

## Checklist

- [x] I have updated the cargo docs to reflect changes made by this PR _(if applicable)_
- [ ] I have updated the `book/` to reflect changes made by this PR _(if applicable)_

### CODE REVIEW

**Comment by ComposableFiB0t** [comment]
Task linked: [CU-2nnhxa9 Remove liquidity](https://app.clickup.com/t/2nnhxa9)

**Comment by ComposableFiB0t** [comment]
Task linked: [CU-2rpv7yc Adapt to new changes in Pablo](https://app.clickup.com/t/2rpv7yc)

**Comment by ComposableFiB0t** [comment]
Task linked: [CU-2qevhzq Tests for fee](https://app.clickup.com/t/2qevhzq)

**Comment by ComposableFiB0t** [comment]
Task linked: [CU-2mkb3vt `ConstantProduct` type of pools](https://app.clickup.com/t/2mkb3vt)

**Comment by ComposableFiB0t** [comment]
Task linked: [CU-2nnhwhf Tests](https://app.clickup.com/t/2nnhwhf)

**Comment by ComposableFiB0t** [comment]
Task linked: [CU-2rh62qe Prop tests for adding/removing liquidity](https://app.clickup.com/t/2rh62qe)

**Comment by ComposableFiB0t** [comment]
Task linked: [CU-2q94zpw Add fee](https://app.clickup.com/t/2q94zpw)

**Comment by ComposableFiB0t** [comment]
Task linked: [CU-2nnhx8y Add liquidity](https://app.clickup.com/t/2nnhx8y)

**Comment by ComposableFiB0t** [comment]
Task linked: [CU-2pj9but Simulate add liquidity](https://app.clickup.com/t/2pj9but)

**Comment by ComposableFiB0t** [comment]
Task linked: [CU-2pcfnf3 Simulate remove liquidity](https://app.clickup.com/t/2pcfnf3)

File: frame/composable-traits/src/dex.rs
Comment (benluelo) [inline frame/composable-traits/src/dex.rs:40]: Consider refactoring this into a more semantic enum? It's difficult to figure out what `true`/`false` mean at the call site

**Review by benluelo — APPROVED** [review]
LGTM, comment is non-critical

File: frame/composable-traits/src/dex.rs
Comment (saruman9) [reply frame/composable-traits/src/dex.rs:40]: I completely agree with you. I will refactor this when I add the asset selection functionality for depositing/withdrawing liquidity (now base asset is used by default). Ok?

Comment (benluelo) [reply frame/composable-traits/src/dex.rs:40]: awesome! 🚀

**Review by kevin-composable — COMMENTED** [review]
Did a first pass through this and everything looks good. I agree with the comment Ben left as well. Is there an argument to be made to have a separate extrinsic for `single_asset_deposits` and `single_asset_withdraws`, rather than the `is_single_deposit` parameter approach?

I am leaving this comment and not approving so it gives time for the other Pablo devs to review as well. I also want to do a second pass through to check the correctness of the math.

**Comment by vimukthi-git** [comment]
@saruman9 @kevin-composable 

I also think things look good here on the first pass. Still need to review a bit more on fee stuff

>Is there an argument to be made to have a separate extrinsic for single_asset_deposits and single_asset_withdraws, rather than the is_single_deposit parameter approach?

This would be good to consider. As flags cause changes to existing code including FE.

Unfortunately since this feature is not a priority for Picasso/Pablo launch, I have to suggest that you maintain a branch of these changes which integrates with the needed pallets from Composable Labs (instead of `main`). We can still review your code before merge to that branch. Just that we want to avoid any new testing/audits/fixes that would be caused by this feature for Pablo before launch. I think keeping things updated in that new branch should be easy given that you guys mostly work on an isolated set of pallets. We will merge these changes later into main when we are ready launch this feature with a runtime upgrade after launch.

cc @KaiserKarel

File: frame/composable-traits/src/dex.rs
Comment (saruman9) [reply frame/composable-traits/src/dex.rs]: Fixed by 9a2720c6005c160cb68ae969e0dd404f6a29479a.

I will update the weights later and recheck the code again, because there are many places where changes have occurred.

**Comment by kevin-composable** [comment]
> Unfortunately since this feature is not a priority for Picasso/Pablo launch, I have to suggest that you maintain a branch of these changes which integrates with the needed pallets from Composable Labs (instead of main). We can still review your code before merge to that branch. Just that we want to avoid any new testing/audits/fixes that would be caused by this feature for Pablo before launch. I think keeping things updated in that new branch should be easy given that you guys mostly work on an isolated set of pallets. We will merge these changes later into main when we are ready launch this feature with a runtime upgrade after launch.

@vimukthi-git sounds good, we will go ahead and make that change. Do you have a specific way you like Pablo's branches to be named?

**Comment by ComposableFiB0t** [comment]
Task linked: [CU-2tkd28t Split the function of removing liquidity](https://app.clickup.com/t/2tkd28t)

**Comment by saruman9** [comment]
To withdraw liquidity in a single asset, I created a separate function (thus got rid of the flag). For the runtime, it is necessary to recalculate the weights, this will be done after the withdrawal of liquidity in a single asset for a pool of the `StableSwap` type is implemented.

I also made minor corrections to `dex-router` in order, among other things, to satisfy the requirements of CI checks.

File: runtime/dali/src/weights/dex_router.rs
Comment (belousm) [inline runtime/dali/src/weights/dex_router.rs:98]: Are you sure that it is the right weight? Or you just chose it for MVP?

Comment (saruman9) [reply runtime/dali/src/weights/dex_router.rs:98]: This file is generated automatically, I inserted these numbers manually for the reason to pass compiler checks (including CI checks). See comments about [`pablo`'s](https://github.com/ComposableFi/composable/blob/84ea1296041a8cb19d9d9caa26f8c8a00283ef86/frame/pablo/src/benchmarking.rs#L154) and [`dex-router`'s](https://github.com/ComposableFi/composable/blob/84ea1296041a8cb19d9d9caa26f8c8a00283ef86/frame/dex-router/src/benchmarking.rs#L261) benchmarks.

File: runtime/dali/src/weights/pablo.rs
Comment (saruman9) [reply runtime/dali/src/weights/pablo.rs:85]: https://github.com/ComposableFi/composable/pull/1416#discussion_r949103651

File: runtime/dali/src/weights/dex_router.rs
Comment (belousm) [reply runtime/dali/src/weights/dex_router.rs:98]: Got it!

**Review by belousm — APPROVED** [review]
Great job! Awesome!

**Review by kevin-composable — COMMENTED** [review]
Everything looks good, however we don't want these changes merged into main

**Pull request closed without merging**
